File: 24000s/24398.py
import sys
sys.setrecursionlimit(10**6)

def select(p, r, q):
    global a
    if p == r:
        return a[p]
    t = partition(p, r)
    k = t - p + 1
    if q < k:
        return select(p, t-1, q)
    elif q == k:
        return a[t]
    else:
        return select(t+1, r, q-k)

def partition(p, r):
    global a, k
    x = a[r]
    i = p - 1
    for j in range(p, r):
        if a[j] <= x:
            i += 1
            k -= 1
            a[i], a[j] = a[j], a[i]
            if k == 0:
                print(min(a[i], a[j]), max(a[i], a[j]))
                exit(0)
    if i + 1 != r:
        k -= 1
        a[i+1], a[r] = a[r], a[i+1]
        if k == 0:
            print(min(a[i+1], a[r]), max(a[i+1], a[r]))
            exit(0)
    return i+1

n, q, k = map(int, input().split())
a = list(map(int, input().split()))
select(0, n-1, q)
print(-1)

File: 24000s/24398.cpp
#include <bits/stdc++.h>
using namespace std;

int n, q, k;
vector<int> a;

int partition(int p, int r) {
    int x = a[r];
    int i = p-1;
    for(int j=p ; j<r ; j++) {
        if(a[j] <= x) {
            i++;
            k--;
            swap(a[i], a[j]);
            if(!k) {
                cout << min(a[i], a[j]) << ' ' << max(a[i], a[j]);
                exit(0);
            }
        }
    }
    if(i+1 != r) {
        k--;
        swap(a[i+1], a[r]);
        if(!k) {
            cout << min(a[i+1], a[r]) << ' ' << max(a[i+1], a[r]);
            exit(0);
        }
    }
    return i+1;
}

int select(int p, int r, int q) {
    if(p == r) return a[p];
    int t = partition(p, r);
    int k = t - p + 1;
    if(q < k) return select(p, t-1, q);
    else if(q == k) return a[t];
    else return select(t+1, r, q-k);
}

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> q >> k;
    for(int i=0 ; i<n ; i++) {
        int t;
        cin >> t;
        a.push_back(t);
    }
    select(0, n-1, q);
    cout << -1;
}